File: public/js/leave_allocation.js
// Dynamic Leave Allocation JavaScript Functions

frappe.provide('aspirehr.leave_allocation');

aspirehr.leave_allocation = {
    
    // Show dynamic leave allocation dialog
    show_allocation_dialog: function() {
        let d = new frappe.ui.Dialog({
            title: __('Dynamic Leave Allocation'),
            fields: [
                {
                    fieldtype: 'Section Break',
                    label: __('Allocation Details')
                },
                {
                    fieldtype: 'Link',
                    fieldname: 'employee',
                    label: __('Employee'),
                    options: 'Employee',
                    description: __('Leave empty to allocate for all active employees')
                },
                {
                    fieldtype: 'Link',
                    fieldname: 'leave_type',
                    label: __('Leave Type'),
                    options: 'Leave Type',
                    description: __('Leave empty to allocate for all active leave types')
                },
                {
                    fieldtype: 'Date',
                    fieldname: 'allocation_date',
                    label: __('Allocation Date'),
                    default: frappe.datetime.get_today(),
                    reqd: 1
                },
                {
                    fieldtype: 'Float',
                    fieldname: 'new_leaves_allocated',
                    label: __('Leaves to Allocate'),
                    reqd: 1,
                    description: __('Number of leaves to allocate')
                },
                {
                    fieldtype: 'Column Break'
                },
                {
                    fieldtype: 'Check',
                    fieldname: 'carry_forward',
                    label: __('Carry Forward Previous Leaves'),
                    description: __('Include unused leaves from previous allocation period')
                },
                {
                    fieldtype: 'Section Break',
                    label: __('Preview & Validation')
                },
                {
                    fieldtype: 'Button',
                    fieldname: 'validate_data',
                    label: __('Validate Allocation Data'),
                    click: function() {
                        aspirehr.leave_allocation.validate_allocation_data(d);
                    }
                },
                {
                    fieldtype: 'HTML',
                    fieldname: 'validation_result',
                    label: __('Validation Result')
                }
            ],
            primary_action_label: __('Allocate Leaves'),
            primary_action: function() {
                aspirehr.leave_allocation.process_allocation(d);
            },
            secondary_action_label: __('Bulk Upload'),
            secondary_action: function() {
                aspirehr.leave_allocation.show_bulk_upload_dialog();
            }
        });
        
        d.show();
    },
    
    // Validate allocation data
    validate_allocation_data: function(dialog) {
        let values = dialog.get_values();
        
        if (!values.allocation_date || !values.new_leaves_allocated) {
            frappe.msgprint(__('Please fill Allocation Date and Leaves to Allocate'));
            return;
        }
        
        frappe.call({
            method: 'aspirehr.api.leave_allocation.validate_leave_allocation_data',
            args: {
                employee: values.employee,
                leave_type: values.leave_type,
                allocation_date: values.allocation_date,
                new_leaves_allocated: values.new_leaves_allocated
            },
            callback: function(r) {
                if (r.message) {
                    let result = r.message;
                    let html = '<div class="validation-result">';
                    
                    if (result.valid) {
                        html += '<div class="alert alert-success"><i class="fa fa-check"></i> Validation Passed</div>';
                    } else {
                        html += '<div class="alert alert-danger"><i class="fa fa-times"></i> Validation Failed</div>';
                    }
                    
                    if (result.errors && result.errors.length > 0) {
                        html += '<div class="alert alert-danger"><strong>Errors:</strong><ul>';
                        result.errors.forEach(error => {
                            html += `<li>${error}</li>`;
                        });
                        html += '</ul></div>';
                    }
                    
                    if (result.warnings && result.warnings.length > 0) {
                        html += '<div class="alert alert-warning"><strong>Warnings:</strong><ul>';
                        result.warnings.forEach(warning => {
                            html += `<li>${warning}</li>`;
                        });
                        html += '</ul></div>';
                    }
                    
                    html += '</div>';
                    
                    dialog.fields_dict.validation_result.$wrapper.html(html);
                }
            }
        });
    },
    
    // Process leave allocation
    process_allocation: function(dialog) {
        let values = dialog.get_values();
        
        if (!values.allocation_date || !values.new_leaves_allocated) {
            frappe.msgprint(__('Please fill all required fields'));
            return;
        }
        
        frappe.confirm(
            __('Are you sure you want to proceed with the leave allocation?'),
            function() {
                frappe.call({
                    method: 'aspirehr.api.leave_allocation.allocate_employee_leaves',
                    args: values,
                    freeze: true,
                    freeze_message: __('Processing leave allocation...'),
                    callback: function(r) {
                        if (r.message && r.message.success) {
                            frappe.msgprint({
                                title: __('Success'),
                                message: r.message.message,
                                indicator: 'green'
                            });
                            
                            dialog.hide();
                            
                            // Show result summary
                            aspirehr.leave_allocation.show_allocation_summary(r.message.data);
                        }
                    }
                });
            }
        );
    },
    
    // Show allocation summary
    show_allocation_summary: function(data) {
        if (!data || data.length === 0) {
            return;
        }
        
        let html = '<div class="allocation-summary">';
        html += `<p><strong>Total Allocations Processed:</strong> ${data.length}</p>`;
        html += '<table class="table table-striped">';
        html += '<thead><tr><th>Employee</th><th>Leave Type</th><th>Leaves Allocated</th><th>Allocation ID</th></tr></thead>';
        html += '<tbody>';
        
        data.forEach(allocation => {
            html += `<tr>
                <td>${allocation.employee}</td>
                <td>${allocation.leave_type}</td>
                <td>${allocation.leaves_allocated}</td>
                <td><a href="/app/leave-allocation/${allocation.allocation}" target="_blank">${allocation.allocation}</a></td>
            </tr>`;
        });
        
        html += '</tbody></table></div>';
        
        frappe.msgprint({
            title: __('Allocation Summary'),
            message: html,
            wide: true
        });
    },
    
    // Show bulk upload dialog
    show_bulk_upload_dialog: function() {
        let d = new frappe.ui.Dialog({
            title: __('Bulk Leave Allocation'),
            fields: [
                {
                    fieldtype: 'Section Break',
                    label: __('Upload CSV File')
                },
                {
                    fieldtype: 'HTML',
                    fieldname: 'upload_help',
                    options: `
                        <div class="alert alert-info">
                            <strong>CSV Format:</strong> employee,leave_type,allocation_date,new_leaves_allocated,carry_forward<br>
                            <strong>Example:</strong><br>
                            EMP-001,Annual Leave,2024-01-01,21,1<br>
                            EMP-002,Sick Leave,2024-01-01,12,0
                        </div>
                    `
                },
                {
                    fieldtype: 'Attach',
                    fieldname: 'csv_file',
                    label: __('CSV File'),
                    reqd: 1
                },
                {
                    fieldtype: 'Section Break',
                    label: __('Or Enter Data Manually')
                },
                {
                    fieldtype: 'Table',
                    fieldname: 'allocation_data',
                    label: __('Allocation Data'),
                    fields: [
                        {
                            fieldtype: 'Link',
                            fieldname: 'employee',
                            label: __('Employee'),
                            options: 'Employee',
                            in_list_view: 1,
                            reqd: 1
                        },
                        {
                            fieldtype: 'Link',
                            fieldname: 'leave_type',
                            label: __('Leave Type'),
                            options: 'Leave Type',
                            in_list_view: 1,
                            reqd: 1
                        },
                        {
                            fieldtype: 'Date',
                            fieldname: 'allocation_date',
                            label: __('Allocation Date'),
                            in_list_view: 1,
                            reqd: 1
                        },
                        {
                            fieldtype: 'Float',
                            fieldname: 'new_leaves_allocated',
                            label: __('Leaves'),
                            in_list_view: 1,
                            reqd: 1
                        },
                        {
                            fieldtype: 'Check',
                            fieldname: 'carry_forward',
                            label: __('Carry Forward'),
                            in_list_view: 1
                        }
                    ]
                }
            ],
            primary_action_label: __('Process Bulk Allocation'),
            primary_action: function() {
                aspirehr.leave_allocation.process_bulk_allocation(d);
            }
        });
        
        d.show();
    },
    
    // Process bulk allocation
    process_bulk_allocation: function(dialog) {
        let values = dialog.get_values();
        let allocation_data = values.allocation_data;
        
        if (!allocation_data || allocation_data.length === 0) {
            frappe.msgprint(__('Please add allocation data'));
            return;
        }
        
        frappe.call({
            method: 'aspirehr.api.leave_allocation.bulk_allocate_leaves',
            args: {
                allocation_data: JSON.stringify(allocation_data)
            },
            freeze: true,
            freeze_message: __('Processing bulk allocation...'),
            callback: function(r) {
                if (r.message && r.message.success) {
                    dialog.hide();
                    
                    let result = r.message;
                    let html = `
                        <div class="bulk-allocation-result">
                            <div class="row">
                                <div class="col-sm-4 text-center">
                                    <div class="alert alert-info">
                                        <h4>${result.summary.total_records}</h4>
                                        <p>Total Records</p>
                                    </div>
                                </div>
                                <div class="col-sm-4 text-center">
                                    <div class="alert alert-success">
                                        <h4>${result.summary.successful}</h4>
                                        <p>Successful</p>
                                    </div>
                                </div>
                                <div class="col-sm-4 text-center">
                                    <div class="alert alert-danger">
                                        <h4>${result.summary.failed}</h4>
                                        <p>Failed</p>
                                    </div>
                                </div>
                            </div>
                    `;
                    
                    if (result.failed && result.failed.length > 0) {
                        html += '<div class="alert alert-warning"><strong>Failed Records:</strong><ul>';
                        result.failed.forEach(failed => {
                            html += `<li>Employee: ${failed.data.employee}, Error: ${failed.error}</li>`;
                        });
                        html += '</ul></div>';
                    }
                    
                    html += '</div>';
                    
                    frappe.msgprint({
                        title: __('Bulk Allocation Result'),
                        message: html,
                        wide: true
                    });
                }
            }
        });
    },
    
    // Show leave adjustment dialog
    show_adjustment_dialog: function() {
        let d = new frappe.ui.Dialog({
            title: __('Adjust Leave Allocation'),
            fields: [
                {
                    fieldtype: 'Link',
                    fieldname: 'employee',
                    label: __('Employee'),
                    options: 'Employee',
                    reqd: 1
                },
                {
                    fieldtype: 'Link',
                    fieldname: 'leave_type',
                    label: __('Leave Type'),
                    options: 'Leave Type',
                    reqd: 1
                },
                {
                    fieldtype: 'Date',
                    fieldname: 'adjustment_date',
                    label: __('Adjustment Date'),
                    default: frappe.datetime.get_today(),
                    reqd: 1
                },
                {
                    fieldtype: 'Float',
                    fieldname: 'adjustment_amount',
                    label: __('Adjustment Amount'),
                    description: __('Positive for addition, negative for deduction'),
                    reqd: 1
                },
                {
                    fieldtype: 'Small Text',
                    fieldname: 'reason',
                    label: __('Reason for Adjustment'),
                    reqd: 1
                }
            ],
            primary_action_label: __('Adjust Allocation'),
            primary_action: function() {
                let values = d.get_values();
                
                frappe.call({
                    method: 'aspirehr.api.leave_allocation.adjust_employee_leave_allocation',
                    args: values,
                    callback: function(r) {
                        if (r.message && r.message.success) {
                            frappe.msgprint({
                                title: __('Success'),
                                message: r.message.message,
                                indicator: 'green'
                            });
                            d.hide();
                        }
                    }
                });
            }
        });
        
        d.show();
    },
    
    // Show leave balance dialog
    show_balance_dialog: function() {
        let d = new frappe.ui.Dialog({
            title: __('Check Leave Balance'),
            fields: [
                {
                    fieldtype: 'Link',
                    fieldname: 'employee',
                    label: __('Employee'),
                    options: 'Employee',
                    reqd: 1
                },
                {
                    fieldtype: 'Link',
                    fieldname: 'leave_type',
                    label: __('Leave Type'),
                    options: 'Leave Type',
                    description: __('Leave empty to show all leave types')
                },
                {
                    fieldtype: 'Date',
                    fieldname: 'as_on_date',
                    label: __('As On Date'),
                    default: frappe.datetime.get_today(),
                    reqd: 1
                },
                {
                    fieldtype: 'Section Break'
                },
                {
                    fieldtype: 'HTML',
                    fieldname: 'balance_result',
                    label: __('Leave Balance')
                }
            ],
            primary_action_label: __('Check Balance'),
            primary_action: function() {
                let values = d.get_values();
                
                frappe.call({
                    method: 'aspirehr.api.leave_allocation.get_employee_leave_balance',
                    args: values,
                    callback: function(r) {
                        if (r.message && r.message.success) {
                            let html = '<div class="leave-balance-result">';
                            
                            if (r.message.balances) {
                                // Multiple leave types
                                html += '<table class="table table-striped">';
                                html += '<thead><tr><th>Leave Type</th><th>Balance</th></tr></thead>';
                                html += '<tbody>';
                                
                                r.message.balances.forEach(balance => {
                                    html += `<tr><td>${balance.leave_type}</td><td>${balance.balance}</td></tr>`;
                                });
                                
                                html += '</tbody></table>';
                            } else {
                                // Single leave type
                                html += `<div class="alert alert-info">
                                    <h4>${r.message.leave_type}: ${r.message.balance} days</h4>
                                    <p>As on ${r.message.as_on_date}</p>
                                </div>`;
                            }
                            
                            html += '</div>';
                            
                            d.fields_dict.balance_result.$wrapper.html(html);
                        }
                    }
                });
            }
        });
        
        d.show();
    }
};

// Add to global namespace for easy access
window.aspirehr_leave_allocation = aspirehr.leave_allocation;

File: api/leave_allocation.py
try:
	import frappe
	from frappe import _
	from frappe.utils import today, getdate, cint
except ImportError:
	# Handle case when frappe is not available (e.g., during development)
	import datetime
	
	class MockFrappe:
		def __init__(self):
			self.session = type('obj', (object,), {'user': 'Administrator'})()
			self.db = type('obj', (object,), {
				'get_value': lambda *args, **kwargs: None,
				'get_all': lambda *args, **kwargs: [],
				'exists': lambda *args, **kwargs: False,
				'commit': lambda: None
			})()
			self.utils = type('obj', (object,), {
				'cint': lambda x: int(x) if x else 0
			})()
		
		def throw(self, msg):
			raise Exception(msg)
		
		def has_permission(self, doctype, ptype="read"):
			return True
		
		def log_error(self, msg):
			print(f"Error: {msg}")
		
		def whitelist(self):
			def decorator(func):
				return func
			return decorator
	
	frappe = MockFrappe()
	
	def _(text):
		return text
	
	def today():
		return datetime.date.today().strftime("%Y-%m-%d")
	
	def getdate(date_str):
		if isinstance(date_str, str):
			return datetime.datetime.strptime(date_str, "%Y-%m-%d").date()
		return date_str
	
	def cint(value):
		return int(value) if value else 0

from aspirehr.human_resources.utils import (
    allocate_leaves_dynamically,
    bulk_leave_allocation,
    adjust_leave_allocation,
    get_leave_balance_on_date
)


@frappe.whitelist()
def allocate_employee_leaves(employee=None, leave_type=None, allocation_date=None, 
                           new_leaves_allocated=None, carry_forward=False):
    """
    API endpoint to dynamically allocate leaves for employees
    
    Args:
        employee (str): Employee ID (optional - if not provided, applies to all active employees)
        leave_type (str): Leave Type (optional - if not provided, applies to all active leave types)
        allocation_date (str): Date for allocation (optional - defaults to today)
        new_leaves_allocated (float): Number of leaves to allocate
        carry_forward (bool): Whether to carry forward unused leaves from previous period
    
    Returns:
        dict: Result with processed allocations and any errors
    """
    
    # Check permissions
    if not frappe.has_permission("Leave Allocation", "create"):
        frappe.throw(_("Not permitted to create Leave Allocations"))
    
    try:
        # Convert string values to appropriate types
        if new_leaves_allocated:
            new_leaves_allocated = float(new_leaves_allocated)
        
        carry_forward = cint(carry_forward)
        
        result = allocate_leaves_dynamically(
            employee=employee,
            leave_type=leave_type, 
            allocation_date=allocation_date,
            new_leaves_allocated=new_leaves_allocated,
            carry_forward=carry_forward
        )
        
        frappe.db.commit()
        
        return {
            "success": True,
            "message": _("Leave allocation completed successfully"),
            "data": result
        }
        
    except Exception as e:
        frappe.log_error(f"Leave allocation API error: {str(e)}")
        frappe.throw(_("Error in leave allocation: {0}").format(str(e)))


@frappe.whitelist()
def bulk_allocate_leaves(allocation_data):
    """
    API endpoint for bulk leave allocation
    
    Args:
        allocation_data (str): JSON string containing list of allocation records
        
    Example allocation_data:
    [
        {
            "employee": "EMP-001",
            "leave_type": "Annual Leave", 
            "allocation_date": "2024-01-01",
            "new_leaves_allocated": 21,
            "carry_forward": true
        }
    ]
    """
    
    # Check permissions
    if not frappe.has_permission("Leave Allocation", "create"):
        frappe.throw(_("Not permitted to create Leave Allocations"))
    
    try:
        # Parse allocation data
        if isinstance(allocation_data, str):
            import json
            allocation_data = json.loads(allocation_data)
        
        result = bulk_leave_allocation(allocation_data)
        
        frappe.db.commit()
        
        return {
            "success": True,
            "message": _("Bulk leave allocation completed"),
            "processed": result["processed"],
            "failed": result["failed"],
            "summary": {
                "total_records": len(allocation_data),
                "successful": len(result["processed"]),
                "failed": len(result["failed"])
            }
        }
        
    except Exception as e:
        frappe.log_error(f"Bulk leave allocation API error: {str(e)}")
        frappe.throw(_("Error in bulk leave allocation: {0}").format(str(e)))


@frappe.whitelist()
def adjust_employee_leave_allocation(employee, leave_type, adjustment_date, 
                                   adjustment_amount, reason=None):
    """
    API endpoint to adjust existing leave allocation
    
    Args:
        employee (str): Employee ID
        leave_type (str): Leave Type
        adjustment_date (str): Date of adjustment
        adjustment_amount (float): Amount to adjust (positive for addition, negative for deduction)
        reason (str): Reason for adjustment (optional)
    """
    
    # Check permissions
    if not frappe.has_permission("Leave Allocation", "write"):
        frappe.throw(_("Not permitted to modify Leave Allocations"))
    
    try:
        adjustment_amount = float(adjustment_amount)
        
        result = adjust_leave_allocation(
            employee=employee,
            leave_type=leave_type,
            adjustment_date=adjustment_date,
            adjustment_amount=adjustment_amount,
            reason=reason
        )
        
        frappe.db.commit()
        
        return {
            "success": True,
            "message": _("Leave allocation adjusted successfully"),
            "allocation": result.name,
            "new_total": result.total_leaves_allocated
        }
        
    except Exception as e:
        frappe.log_error(f"Leave adjustment API error: {str(e)}")
        frappe.throw(_("Error in leave allocation adjustment: {0}").format(str(e)))


@frappe.whitelist()
def get_employee_leave_balance(employee, leave_type=None, as_on_date=None):
    """
    API endpoint to get leave balance for an employee
    
    Args:
        employee (str): Employee ID
        leave_type (str): Leave Type (optional - if not provided, returns all leave types)
        as_on_date (str): Date to check balance (optional - defaults to today)
    """
    
    # Check permissions
    if not frappe.has_permission("Leave Allocation", "read"):
        frappe.throw(_("Not permitted to view Leave Allocations"))
    
    try:
        if not as_on_date:
            as_on_date = today()
        
        if leave_type:
            # Get balance for specific leave type
            balance = get_leave_balance_on_date(employee, leave_type, as_on_date)
            return {
                "success": True,
                "employee": employee,
                "leave_type": leave_type,
                "balance": balance,
                "as_on_date": as_on_date
            }
        else:
            # Get balance for all leave types
            leave_types = frappe.get_all("Leave Type", 
                filters={"is_active": 1}, 
                fields=["name"]
            )
            
            balances = []
            for lt in leave_types:
                balance = get_leave_balance_on_date(employee, lt.name, as_on_date)
                balances.append({
                    "leave_type": lt.name,
                    "balance": balance
                })
            
            return {
                "success": True,
                "employee": employee,
                "balances": balances,
                "as_on_date": as_on_date
            }
            
    except Exception as e:
        frappe.log_error(f"Leave balance API error: {str(e)}")
        frappe.throw(_("Error getting leave balance: {0}").format(str(e)))


@frappe.whitelist()
def get_leave_allocation_history(employee, leave_type=None, from_date=None, to_date=None):
    """
    API endpoint to get leave allocation history for an employee
    
    Args:
        employee (str): Employee ID
        leave_type (str): Leave Type (optional)
        from_date (str): Start date for history (optional)
        to_date (str): End date for history (optional)
    """
    
    # Check permissions
    if not frappe.has_permission("Leave Allocation", "read"):
        frappe.throw(_("Not permitted to view Leave Allocations"))
    
    try:
        filters = {"employee": employee, "docstatus": 1}
        
        if leave_type:
            filters["leave_type"] = leave_type
            
        if from_date:
            filters["from_date"] = [">=", from_date]
            
        if to_date:
            filters["to_date"] = ["<=", to_date]
        
        allocations = frappe.get_all("Leave Allocation",
            filters=filters,
            fields=["name", "leave_type", "from_date", "to_date", 
                   "new_leaves_allocated", "carry_forwarded_leaves", 
                   "total_leaves_allocated", "creation", "modified"],
            order_by="from_date desc"
        )
        
        return {
            "success": True,
            "employee": employee,
            "allocations": allocations
        }
        
    except Exception as e:
        frappe.log_error(f"Leave allocation history API error: {str(e)}")
        frappe.throw(_("Error getting leave allocation history: {0}").format(str(e)))


@frappe.whitelist()
def validate_leave_allocation_data(employee, leave_type, allocation_date, new_leaves_allocated):
    """
    API endpoint to validate leave allocation data before processing
    
    Args:
        employee (str): Employee ID
        leave_type (str): Leave Type
        allocation_date (str): Allocation date
        new_leaves_allocated (float): Number of leaves to allocate
    """
    
    try:
        errors = []
        warnings = []
        
        # Validate employee
        if not frappe.db.exists("Employee", employee):
            errors.append(f"Employee {employee} does not exist")
        else:
            emp_status = frappe.db.get_value("Employee", employee, "status")
            if emp_status != "Active":
                warnings.append(f"Employee {employee} is not active (Status: {emp_status})")
        
        # Validate leave type
        if not frappe.db.exists("Leave Type", leave_type):
            errors.append(f"Leave Type {leave_type} does not exist")
        else:
            lt_active = frappe.db.get_value("Leave Type", leave_type, "is_active")
            if not lt_active:
                warnings.append(f"Leave Type {leave_type} is not active")
        
        # Validate allocation amount
        if float(new_leaves_allocated) < 0:
            errors.append("Leave allocation cannot be negative")
        
        # Check for existing allocation
        allocation_date = getdate(allocation_date)
        
        existing = frappe.db.get_value("Leave Allocation", {
            "employee": employee,
            "leave_type": leave_type,
            "from_date": ["<=", allocation_date],
            "to_date": [">=", allocation_date],
            "docstatus": ["!=", 2]
        })
        
        if existing:
            warnings.append(f"An allocation already exists for this period: {existing}")
        
        return {
            "valid": len(errors) == 0,
            "errors": errors,
            "warnings": warnings
        }
        
    except Exception as e:
        frappe.log_error(f"Leave allocation validation error: {str(e)}")
        return {
            "valid": False,
            "errors": [f"Validation error: {str(e)}"],
            "warnings": []
        }
